### PR TITLE
Ncf docs

### PR DESCRIPTION
**Pull request by miguelgfierro**
### Description

Updated docstrings. 

There is one change that I'm not sure if it will change the results of the [benchmark](https://github.com/Microsoft/Recommenders/blob/master/benchmarks/movielens.ipynb), the results of NCF are very good, not sure because of this. The original code had a flag for `implicit=True` by default, which was inaccessible, because it was in a private function. I added it to the constructor, so now we can access it. The problem is that this flag actually is doing this: `If true, set rating > 0 to rating = 1`. However, Movielens is not implicit, it is explicit, and here we are treating it like implicit. Any feedback about this @AaronHeee?

### Related Issues

https://github.com/Microsoft/Recommenders/issues/597

### Checklist:

- [ ] My code follows the code style of this project, as detailed in our [contribution guidelines](../CONTRIBUTING.md).
- [ ] I have added tests.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by loomlike** [comment]
Here are my thought:

This is because our ranking metrics don't care about if user will like the movie or not,
while our models (w/ explicit feedback) learn like-or-not.
(note, our implementation users binary relevancy score).
 
E.g. Rating data of a user: `train = [5, 1, 1, x, x, x, x]`,  `test = [1, 1, 1, x, x, x, x, x]`,
x means the user did not watched (or not-rated).

And now let's say we have a **perfect model** M. Because M is so good, it predicts the future,
which will be `test = [1, 1, 1, 3, 4, 5, 2, 2]`.
You can see, our ranking metrics (w/ k=3) will give *F* (F for fail) to our model because the movies user watched (rated) in our test data will not be in top-3.
On the other hand, if the model learned 'watch-or-not', it would get higher ranking metrics.

**Comment by loomlike** [comment]
Similarly, when I tried ALS w/ `implicit feedback = True` on Movielens, it performed better than those of ALS w/ explicit-feedback-setting.

**Comment by yueguoguo** [comment]
This is related to the "negative sampling" feature in the implementation. To account for the negative (implicit) feedback, @AaronHeee has sampled 1 negative sample in the data, and this is from the movies that have been watched by the users. This is what people usually do in processing the Movielens data for implicit feedback typed settings. @AaronHeee correct me if there is anything. 

I think @loomlike comment explained well about "implicit feedback does not predict the preference (e.g., rating) but the likelihood of interaction (in our context, watch a movie)". So in the ranking metrics, the relevant items, i.e., the list of items that will be very likely to be watched by a user, will be evaluated against the actual items.

**Comment by AaronHeee** [comment]
> This is related to the "negative sampling" feature in the implementation. To account for the negative (implicit) feedback, @AaronHeee has sampled 1 negative sample in the data, and this is from the movies that have been watched by the users. This is what people usually do in processing the Movielens data for implicit feedback typed settings. @AaronHeee correct me if there is anything.

That's right, and you can view the number of negative sample as a hyper-parameter, tune it for the best performance.

File: reco_utils/recommender/ncf/ncf_singlenode.py [hunk 1]
```diff
@@ -204,40 +219,46 @@ def _create_model(self,):
...
     def load(self, gmf_dir=None, mlp_dir=None, neumf_dir=None, alpha=0.5):
-        """ load model parameters for further use.
-            GMF model --> load parameters in `gmf_dir`
-            MLP model --> load parameters in `mlp_dir`
-            NeuMF model --> load parameters in `neumf_dir` or in `gmf_dir` and `mlp_dir`
-            Args:
-                gmf_dir, mlp_dir, neumf_dir ( str or None ): model parameters directory name
-            Returns:
-                load parameters in this model
+        """Load model parameters for further use.
```
Comment (gramhagen) [inline reco_utils/recommender/ncf/ncf_singlenode.py]: it seems like it would be better to have a static method to load the model, where the model type is specified, and a single directory is provided. even if it's kept as a class method which uses model_type defined, I still think there should only be a single directory given.

Comment (miguelgfierro) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: I was looking how other libs do the loading and saving, in LightGBM they do:
```
lgb_model.save_model(save_file)
loaded_model = lgb.Booster(model_file=save_file)
```
Are you thinking on something like this or different?

Comment (gramhagen) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: yeah, I think there are two straightforward ways with the current implementation:
```
old_model.save(model_dir=model_dir)
// keep load as an instance method
model = NCF(model_type="NeuMF").load(model_dir=model_dir)
... or ...
// make load a class method
model = NCF.load(model_type="NeuMF", model_dir=model_dir)
```
It seems like less work to go with option 1, plus it will let you modify the model parameters if needed and just reload the weights (I'm assuming that's what tf.train.Saver().restore does anyway)

either way the model_dir variable doesn't really need to be tied to the type of model

File: reco_utils/recommender/ncf/dataset.py [hunk 2]
```diff
@@ -26,12 +24,25 @@ def __init__(
...
+            train (pd.DataFrame): Training data with at least columns (col_user, col_item, col_rating).
+            test (pd.DataFrame): Test data with at least columns (col_user, col_item, col_rating). test can be None, 
+                if so, we only process the training data.
+            n_neg (int): Number of negative samples for training set.
+            n_neg_test (int): Number of negative samples for test set.
+            col_user (str): User column name.
+            col_item (str): Item column name.
+            col_rating (str): Rating column name. 
+            col_timestamp (str): Timestamp column name.
+            implicit (bool): If true, set rating > 0 to rating = 1.
```
Comment (anargyri) [inline reco_utils/recommender/ncf/dataset.py]: Do you mean "If true, set rating > 0 else set rating = 1." ?

**Comment by anargyri** [comment]
@miguelgfierro I think there is an issue that relates to the benchmark more generally and not only to NCF. There are a couple of instances where it is not explicit what happens and also it is not explicit that we don't violate any machine learning "firewall". Specifically, in line
`train, test = prepare_metrics_data.get(algo, lambda x,y:(x,y))(df_train, df_test)`
one would be more confident if we separated this in two lines, one with `df_train` only and one with `df_test`, unless there is a good reason. 
And line 
`top_k_scores, time_ranking = ranking_predictor[algo](model, test, train)`
should not have `test` normally. What happens here implicitly is not transparent to the reader (and indeed someone may find it suspect).

**Comment by miguelgfierro** [comment]
@anargyri, I know what you mean, the problem is that it is very difficult to have a common interface for all the algos. I struggle a lot when creating the interfaces, and had to do several passes. Many times it is due to how the different functions are implemented.

About the other comment, do you propose something like:
```
train = prepare_metrics_train_set.get(algo, lambda x:x)(df_train)
test = prepare_metrics_test_set.get(algo, lambda  x:x)(df_test)
```

**Comment by anargyri** [comment]
> @anargyri, I know what you mean, the problem is that it is very difficult to have a common interface for all the algos. I struggled a lot when creating the interfaces, and had to do several passes. Many times it is due to how the different functions are implemented.
> 
> About the other comment, do you propose something like:
> 
> ```
> train = prepare_metrics_train_set.get(algo, lambda x:x)(df_train)
> test = prepare_metrics_test_set.get(algo, lambda  x:x)(df_test)
> ```

Yes, this is exactly what I propose.
About the ranking_prediction, which of the algorithms requires the test set? This would be an issue with the design of the specific algorithm, and should be consistent with the others.

**Comment by miguelgfierro** [comment]
> About the ranking_prediction, which of the algorithms requires the test set? This would be an issue with the design of the specific algorithm, and should be consistent with the others.

bufff, there is a mess here. In SAR for example, we are using the users in the test set to recommend items, whereas in  SVD we are using the users in the train set. I think this need a deep review I created a new issue to address this https://github.com/Microsoft/Recommenders/issues/768

**Comment by anargyri** [comment]
> > About the ranking_prediction, which of the algorithms requires the test set? This would be an issue with the design of the specific algorithm, and should be consistent with the others.
> 
> bufff, there is a mess here. In SAR for example, we are using the users in the test set to recommend items, whereas in SVD we are using the users in the train set. I think this need a deep review I created a new issue to address this #768

Totally agree that we should strive to make the different algos consistent to a common syntax.
However, the SAR method accepts any dataframe you give it, it doesn't need to be the test data set. So for the benchmark you just need to pass a dataframe that contains the users for which you want recommendations. You could compute this dataframe on a separate line.

**Comment by miguelgfierro** [comment]
> About the other comment, do you propose something like:
> train = prepare_metrics_train_set.get(algo, lambda x:x)(df_train)
> test = prepare_metrics_test_set.get(algo, lambda  x:x)(df_test)
> Yes, this is exactly what I propose.

I think in fact that in the ranking metrics the input should be the test set, not the train set. Basically, the test set represent the users that we want to applicate recommendations on. Therefore, after we fix https://github.com/Microsoft/Recommenders/issues/768, there should be no need for `prepare_metrics_train_set`. I think we should fix this issue before changing the code in prepare_metrics. Would this make sense?

File: reco_utils/recommender/ncf/dataset.py (hunk 2 above)
Comment (miguelgfierro) [reply reco_utils/recommender/ncf/dataset.py]: changed

File: reco_utils/recommender/ncf/ncf_singlenode.py (hunk 1 above)
Comment (miguelgfierro) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: about this, I´ve been looking at how they save the models in ligthgbm https://github.com/Microsoft/LightGBM/blob/master/examples/python-guide/advanced_example.py#L83 and sklearn https://machinelearningmastery.com/save-load-machine-learning-models-python-scikit-learn/. Basically, whatever they load, it has all the parameters the algo needs. Right now, whatever we use the current form or the static method we need to define the parameters

```
model = NCF (
    n_users=data.n_users, 
    n_items=data.n_items,
    model_type="NeuMF",
    n_factors=4,
    layer_sizes=[16,8,4],
    n_epochs=EPOCHS,
    batch_size=BATCH_SIZE,
    learning_rate=1e-3,
    verbose=10,
)
model.load(gmf_dir=".pretrain/GMF", mlp_dir=".pretrain/MLP", alpha=0.5)
```
or 
```
model = NCF (
    n_users=data.n_users, 
    n_items=data.n_items,
    model_type="NeuMF",
    n_factors=4,
    layer_sizes=[16,8,4],
    n_epochs=EPOCHS,
    batch_size=BATCH_SIZE,
    learning_rate=1e-3,
    verbose=10,
).load(gmf_dir=".pretrain/GMF", mlp_dir=".pretrain/MLP", alpha=0.5)
```

I think what we should do is this:

```
model = function_or_class_to_load(gmf_dir=".pretrain/GMF", mlp_dir=".pretrain/MLP", alpha=0.5)
```
which includes all the parameters.

Would this make sense?

**Comment by loomlike** [comment]
I have some changes on ncf for `seed` setting. I will wait till this PR get merged and then work on that.

**Comment by miguelgfierro** [comment]
> I have some changes on ncf for seed setting. I will wait till this PR get merged and then work on that.

Where you able to make it repeatable?

File: reco_utils/recommender/ncf/ncf_singlenode.py (hunk 1 above)
Comment (gramhagen) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: that would be great, except for the part where there are multiple directories in the signature. Are both needed for some reason? I was suggesting there just be single directory input on where to look up the model. If we must know they type of model before we access that directory, then it should be a input parameter.

Comment (miguelgfierro) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: both are needed in the case of NeuMF, because it is a mix of GMF and MLP

Comment (gramhagen) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: ah i see, ok, that sounds good then

Comment (miguelgfierro) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: ok I created an issue about this: https://github.com/Microsoft/Recommenders/issues/771. Apart from this, is there anything missing on this PR?

**Comment by yueguoguo** [comment]
The evaluation method takes both train and test as inputs and do the join of the two dataframes, and then calculate the metrics. So even if the input of evaluator is the training set, it still looks at the test data for the users and then do the calculation. 

Does this relate to your discussion above? @miguelgfierro @anargyri

**Comment by anargyri** [comment]
> The evaluation method takes both train and test as inputs and do the join of the two dataframes, and then calculate the metrics. So even if the input of evaluator is the training set, it still looks at the test data for the users and then do the calculation.
> 
> Does this relate to your discussion above? @miguelgfierro @anargyri

No, we were talking about prepare_metrics utilities, which do only data prep (hence there can be separate methods for training and test data sets).
